cishouseholds/validate_class.py

```diff
@@ -0,0 +1,120 @@
+            self.execute_check(check["function"], check["error_message"], column_name, list(method.values())[0])
+
+    def validate(self, operations):
+        for method, params in operations.items():
+            self.execute_check(self.functions[method]["function"], self.functions[method]["error_message"], **params)
+
+    def validate_udl(self, logic, error_message):
+        self.execute_check(logic, error_message)
+
+    def execute_check(self, check, error_message, *params, **kwargs):
+        if callable(check):
+            check, error_message = check(error_message, *params, **kwargs)
+
+        self.dataframe = self.dataframe.withColumn(
+            self.error_column,
+            F.when(~check, F.array_union(F.col(self.error_column), F.array(F.lit(error_message)))).otherwise(
+                F.col(self.error_column)
+            ),
+        )
+
+    @staticmethod
+    def not_null(error_message, column_list):  # works in validate and validate_column
+        error_message = error_message.format(", ".join(column_list))
+        return (~F.array_contains(F.array(*column_list), None)), error_message
+
+    @staticmethod
+    def contains(error_message, column_name, contains):
+        error_message = error_message.format(column_name, contains)
+        return F.col(column_name).rlike(contains), error_message
+
+    @staticmethod
+    def isin(error_message, column_name, options):
+        error_message = error_message.format(column_name, ", ".join(options))
+        return F.col(column_name).isin(options), error_message
+
+    @staticmethod
+    def between(error_message, column_name, range):
+        lb = (
+            (F.col(column_name) >= range["lower_bound"]["value"])
+            if range["lower_bound"]["inclusive"]
+            else (F.col(column_name) > range["lower_bound"]["value"])
+        )
+        ub = (
+            (F.col(column_name) <= range["upper_bound"]["value"])
+            if range["upper_bound"]["inclusive"]
+            else (F.col(column_name) < range["upper_bound"]["value"])
+        )
+        error_message = error_message.format(column_name, *range)
+        return lb & ub, error_message
+
+    # Non column wise functions
+    @staticmethod
+    def duplicated(error_message, column_list):
+        window = Window.partitionBy(*column_list)
+        error_message = error_message.format(", ".join(column_list))
+        return F.when(F.sum(F.lit(1)).over(window) == 1, True).otherwise(False), error_message
+
+    @staticmethod
+    def valid_vaccination(error_message, visit_type_column, check_columns):
+        return (F.col(visit_type_column) != "First Visit") | (
+            ~F.array_contains(F.array(*check_columns), None)
+        ), error_message
+
+
+# calls
+
+calls = {
+    "visit_date": {
+        "between": {
+            "lower_bound": {"inclusive": True, "value": "26/04/2020"},
+            "upper_bound": {"inclusive": True, "value": F.col("file_date")},
```

Comment:
As an update for the requirement here, please could you change this to be the source file date + 1

Does this reference an real column currently? We might need to derive this one as a dependency. I think an existing derivation contains some code to extract the date from the filename, so a dedicated column could be created.